### ISSUE TITLE
Memory sampler shows impossibly large memory allocation when using for each loops in Java 12

### ISSUE THREAD

**Issue by ecen** [oracle/visualvm#140]
**Describe the bug**
In Java 12, the Memory sampler shows impossibly large memory allocation when sampling an application which is continually making use of the java for each loop.

**To Reproduce**
1. Create the following java program:
```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {
    static List<String> list = new ArrayList<>(Arrays.asList("a"));

    public static void main(String[] args) {
        while (true){
            for(String s:list){

            }
        }
    }
}
```
2. In VisualVM, click the Main process the the applications view.
3. Go to the Sampler tab.
4. Click the Memory button.
5. Go to the "Per thread allocations" tab.
6. See that the "main" thread is allocating alot of memory. (4 GB/s in my case)
![4gbs](https://user-images.githubusercontent.com/10086779/60976273-aba3b800-a32d-11e9-98da-e4298ce98f85.png)

**Expected behavior**
The memory allocation per second should be 0 B/s.

Note that everything works as expected (the allocation is 0 B/s) when using a regular for loop like so:
```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {
    static List<String> list = new ArrayList<>(Arrays.asList("a"));

    public static void main(String[] args) {
        while (true){
            for (int i = 0; i < list.size(); i++) {
                String s = list.get(i);

            }
        }
    }
}
```
It is also 0 B/s when using Java 8.

![0bps](https://user-images.githubusercontent.com/10086779/60976329-c4ac6900-a32d-11e9-9bfb-d1f512941dcf.png)

**VisualVM log**
[messages.log](https://github.com/oracle/visualvm/files/3377940/messages.log) from the Arch Linux machine.

**Desktop (please complete the following information):**
 - OS 1: Linux, Arch Linux
 - OS 2: Linux, Manjaro 18.0.4 Illyria
 - JDK version: openjdk 12.0.1 2019-04-16
 - Version: 1.4.3

**Comment by thurka** [comment]
I don't see anything wrong in your example. VisualVM just shows you what is going on. Can you please explain, why 'Memory sampler shows impossibly large memory allocation'? and why 'memory allocation per second should be 0 B/s.' when using for each loop?

**Comment by ecen** [comment]
I would expect the for each loop to allocate no more memory than the regular for loop. Is this incorrect?

The allocation is also shown to be 4 GB per second, which in the screenshot totals to 250 GB of allocated memory. That much space is not available in either my RAM or my SSD. Shouldn't these values reflect the amount of memory that is currently allocated/being allocated?

In addition, in Java 8, there is no ongoing memory allocation using the for each loop (but it is in Java 12).

**Comment by thurka** [comment]
> I would expect the for each loop to allocate no more memory than the regular for loop (which is not show to allocate much memory). Is this incorrect?

Yes, this is incorrect. You clearly see in VisualVM that your example allocates a lot of objects. In fact,  if you can just switch from 'Per thread allocation' to 'Heap histogram' view. You will find ourself, what is going on.

> The allocation is also shown to be 4 GB per second, which in the screenshot totals to 250 GB of allocated memory. That much space is not available in either my RAM or my SSD. Shouldn't these values reflect the amount of memory that is currently allocated/being allocated?

No, there is a difference between allocated and live objects. From your description, it looks to me that you think allocated objects are live objects. 'Allocated Bytes' column is a sum of all objects allocated by appropriate thread - it only keeps growing.

> In addition, in Java 8, there is no ongoing memory allocation using the for each loop (but it is in Java 12).

I don't see this behaviour. For each loop allocates objects in JDK 12 and also in JDK 8 when I tested your example.

**Comment by ecen** [comment]
I went back and tested the same examples with JDK 8 again, and it did indeed show the same behaviour as JDK 12, I must have been mistaken before.

I hadn't thought of looking much at the heap histogram, since I am really using Visual VM to try to account for memory usage in an opengl/lwjgl program that seems to be allocating memory off-heap. This allocation seems to show up in the per-thread view, but not in the heap histogram, so I had disregarded that view somewhat.

Your explanations have made things much clearer in regards to Visual VM however. Thank you! It seems this was all just an error on my part, so I'm closing this issue.
